### ISSUE TITLE
nvFuser's RNG agrees with PyTorch only on `(numel - 1) // 4 + 1` elements

### ISSUE THREAD

**Issue by IvanYashchuk** [NVIDIA/Fuser#557]
Am I correct that nvFuser and PyTorch use the same variant of Philox RNG? 
nvFuser also shares the RNG state with PyTorch so some users might expect the same values to be generated. Is this a reasonable expectation and something we can promise (as long as PyTorch keeps using the same generator as we do)?

I tried exploring whether we're generating the same sequences as PyTorch, and we are not. But there's a pattern that the results match on the `(total_number_of_elements - 1) // 4 + 1` number of elements. So for a size 1 tensor - 1 element is matching, size 5 -> 2 matches, size 9 -> 3 matches, size 13 -> 4 matches, and so on. What's going on here? Can we match PyTorch's behavior or can we make changes to PyTorch so that it matches nvFuser's behavior?

Here's a script that I used:
```py
from nvfuser import FusionDefinition, DataType
import torch
from itertools import product

shape = (13,)
device = "cuda"

# Now let's try what sequence nvfuser will generate
with FusionDefinition() as fd:
    min = fd.define_scalar(0.0)
    max = fd.define_scalar(1.0)
    u = fd.ops.uniform(min, max, [fd.define_scalar(s) for s in shape])
    fd.add_output(u)

# PyTorch Eager and nvFuser use different threading mapping so the position of
# the elements in the output tensor will be different. However, the sorted values
# can be the same.
torch.manual_seed(0)
pt_result = torch.rand(shape, device=device).ravel().sort().values

torch.manual_seed(0)
nvfuser_result = fd.execute([])[0].ravel().sort().values

# count how many nvfuser elements are the same as the pytorch elements
count = sum(x == y for x, y in product(pt_result, nvfuser_result)).item()
print(f"nvfuser and pytorch agree on {count} elements out of {len(pt_result)}")
```

**Comment by zasdfgbnm** [comment]
Your understanding that nvFuser is using the same philox engine as PyTorch, because PyTorch uses curand, and nvFuser's philox engine is copy-pasted from curand. If we do want the nvFuser and PyTorch to generate identical random numbers, I think it makes more sense to modify PyTorch. nvFuser's random number only depend on the logical index of an item in the tensor, not on stride order and schedule. And PyTorch's random number depends on schedule and stride order. For example:
```python
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x7f9da0af4a70>
>>> torch.empty(5,5, device='cuda').random_()
tensor([[ 2615407.,  4527453.,  6411889., 11078092.,  2326039.],
        [16159553.,  4087101., 16538380., 10655256., 14334289.],
        [ 4935680.,  4310594., 12760230., 13040119.,  9751177.],
        [10934575.,  1371688.,  5018307.,  6488175., 10309028.],
        [ 7707908., 12480956., 12017919., 13795622.,  3160016.]],
       device='cuda:0')
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x7f9da0af4a70>
>>> torch.empty(5,5, device='cuda').t().random_()
tensor([[ 2615407., 16159553.,  4935680., 10934575.,  7707908.],
        [ 4527453.,  4087101.,  4310594.,  1371688., 12480956.],
        [ 6411889., 16538380., 12760230.,  5018307., 12017919.],
        [11078092., 10655256., 13040119.,  6488175., 13795622.],
        [ 2326039., 14334289.,  9751177., 10309028.,  3160016.]],
       device='cuda:0')
```

**Comment by IvanYashchuk** [comment]
Before introducing changes to the logical index we need to identify what's causing the discrepancy in the generated values if we are really using the same generator. Note that in the example I have sorted the values to address any potential variations in threading.
It seems that we're using the offset differently since the number of discrepancies strictly follows the formula `(total_number_of_elements - 1) // 4 + 1`

**Comment by zasdfgbnm** [comment]
nvFuser uses exactly one offset per random op: https://github.com/NVIDIA/Fuser/blob/c39ebf29a602fdbdf02a801e30728ddf8645fb1f/csrc/device_lower/lower2device.cpp#L208
PyTorch uses multiple offsets per each random op, depending on the tensor size:
https://github.com/pytorch/pytorch/blob/bfd995f0d6bf87262613b5e89d871832ca9e9938/aten/src/ATen/native/cuda/DistributionTemplates.h#L59

**Comment by zasdfgbnm** [comment]
Philox is a ternary function `philox(seed, subsequence, offset) -> pseudorandom number`. nvFuser maps the logical linear index of the tensor to `subsequence`, and the RNGOp id to `offset`. PyTorch maps CUDA threads to `subsequence`, and `(RNGOp id, item id for this thread)` to `offset`.

**Comment by IvanYashchuk** [comment]
Thank you for the explanations, it's all clear!

One more question, what's the purpose of `rng_component`?
https://github.com/NVIDIA/Fuser/blob/c39ebf29a602fdbdf02a801e30728ddf8645fb1f/csrc/codegen.cpp#L710-L711
Why is this function should depend on the linear index?
https://github.com/NVIDIA/Fuser/blob/c39ebf29a602fdbdf02a801e30728ddf8645fb1f/runtime/random_numbers.cu#L74-L76
With the following diff, I get sorted random numbers from nvFuser exactly matching with PyTorch:
```diff
diff --git a/csrc/codegen.cpp b/csrc/codegen.cpp
index 267670c..de4993b 100644
--- a/csrc/codegen.cpp
+++ b/csrc/codegen.cpp
@@ -710,7 +710,7 @@ class CudaKernelGenerator : private kir::ConstIrVisitor {
     indent() << "nvfuser_index_t rng_subseq" << rop->name() << " = linear_index"
              << rop->name() << " / " << multiple << ";\n";
     indent() << "nvfuser_index_t rng_component" << rop->name()
-             << " = linear_index" << rop->name() << " % " << multiple << ";\n";
+             << " = 0;\n";//linear_index" << rop->name() << " % " << multiple << ";\n";
     indent() << "nvfuser_index_t rng_offset" << rop->name() << " = "
              << rop->getRNGOffset() << ";\n";
     indent() << "if (rng_subseq != rng_subseq" << rop->name()
```

**Comment by IvanYashchuk** [comment]
Hm, PyTorch just drops 3 out of 4 Philox's generated values, but we are using them with the help of `rng_component`.

**Comment by zasdfgbnm** [comment]
I don't think PyTorch discard the rest 3, see https://github.com/pytorch/pytorch/blob/bfd995f0d6bf87262613b5e89d871832ca9e9938/aten/src/ATen/native/cuda/DistributionTemplates.h#L87
where there is a `(&rand.x)[ii]` where `ii` can be 0, 1, 2, 3,

**Comment by IvanYashchuk** [comment]
Right right, and `ii` is more than 0 only for large tensors when the grid-stride loop kicks in, and other than 0th Philox values are beginning to be used then. `ii` is _independent_ of the linear index there.
Whereas in our case, it depends on the linear index and so we're still always discarding 3/4 values for each thread but which value stays is decided based on the linear index.

Do we generate a grid-stride loop when there are not enough threads to fill in the whole tensor?

**Comment by kevinstephano** [comment]
I thought we wanted Pytorch's implementation to be agnostic of kernel blocking and Pytorch eager mode kennels need to be changed in order for that to be true whereas nvFuser already implemented a kernel blocking agnostic implementation.  Therefore, we understand the implementations don't match and it is annoying for debug but the path forward is not for nvFuser to implement Pytorch's specific kernel blocking for RNG operations.

**Comment by zasdfgbnm** [comment]
> I thought we wanted Pytorch's implementation to be agnostic of kernel blocking and Pytorch eager mode kennels need to be changed in order for that to be true whereas nvFuser already implemented a kernel blocking agnostic implementation. Therefore, we understand the implementations don't match and it is annoying for debug but the path forward is not for nvFuser to implement Pytorch's specific kernel blocking for RNG operations.

My understanding matches yours. So there is no actionable on the nvFuser side, but there's an actionable on PyTorch? I am honestly not sure about if we should go and change PyTorch, or just live with this mismatch.

**Comment by jacobhinkle** [comment]
@zasdfgbnm @kevinstephano @IvanYashchuk I think this is safe to close as we do not plan to change the nvFuser implementation. If something changes then we should please reopen or file a new issue.
